doc/reference/trd-hil-design.md

```diff
@@ -0,0 +1,728 @@
+`Ok` result means there will be a callback in response to this call,
+and an `Err` result means there will not be a callback in response to
+*this* call. Note that it is possible for an `Err` result to be
+returned yet there will be a callback in the future.  This depends on
+which `ErrorCode` is passed.  A common calling pattern is for a trait
+to return `ErrorCode::BUSY` if there is already an operation pending
+and a callback will be issued. This error code is unique in this way:
+the general rule is that `Ok` means there will be a callback in
+response to this call, `Err` with `ErrorCode::BUSY` means this call
+did not start a new operation but there will be a callback in response
+to a prior call, and any other `Err` means there will not be a
+callback.
+
+
+Rule 4: Return Passed Buffers in Error Results
+===============================
+
+Consider this method:
+
+```rust
+// Anti-pattern: caller cannot regain buf on an error
+fn send(&self, buf: &'static mut [u8]) -> Result<(), ErrorCode>;
+```
+
+This method is for a split-phase call: there is a corresponding
+completion callback that passes the buffer back:
+
+```rust
+fn send_done(&self, buf: &'static mut[u8]);
+```
+
+
+The `send` method follows Rule 2: it returns a synchronous error. But
+suppose that calling it returns an `Err(ErrorCode)`: what happens to
+the buffer?
+
+Rust's ownership rules mean that the caller can't still hold the reference:
+it passed the reference to the implementer of `send`. But since the
+operation did not succeed, the caller does not expect a callback. Forcing
+the callee to issue a callback on a failed operation typically forces it
+to include an alarm or other timer. Following Rule 1 means it can't do
+so synchronously, so it needs an asynchronous event to invoke the callback
+from. This leads to every implementer of the HIL requiring an alarm or
+timer, which use RAM, has more complex logic, and makes initialization more
+complex.
+
+As a result, in the above interface, if there is an error on `send`, the buffer
+is lost. It's passed into the callee, but the callee
+has no way to pass it back.
+
+If a split-phase operation takes a reference to a buffer as a parameter, it
+should return a reference to a buffer in the `Err` case:
+
+```rust
+fn send(&self, buf: &'static mut [u8]) -> Result<(), (ErrorCode, &'static mut [u8])>;
+```
+
+Before Tock transitioned to using `Result`, this calling pattern was typically
+implemented with an `Option`:
+
+
+```rust
+fn send(&self, buf: &'static mut [u8]) -> (ReturnCode, Option<&'static mut [u8]>);
+```
+
+In this approach, when the `ReturnCode` is `SUCCESS`, the `Option` is always supposed
+to be `None`; it the `ReturnCode` has an error value, the `Option` contains the passed
+buffer. This invariant, however, cannot be checked. Transitioning to using `Result`
+both makes Tock more in line with standard Rust code and enforces the invariant.
+
+
+Rule 5: Always Pass a Mutable Reference to Buffers
```

Comment:
Musing: I wonder if we should have some sort of `UniqueROBuffer` that provided immutable access to a reference, but which is guaranteed to be the only borrow of a given buffer. Something like:

```rust
pub struct UniqueROBuffer { buffer: &'static mut [u8] }

impl From<&'static mut [u8]> for UniqueROBuffer {
    fn from(buffer: &'static mut [u8]) -> UniqueROBuffer { UniqueROBuffer { buffer } }
}

impl From<UniqueROBuffer> for &'static mut [u8] {
    fn from(buffer: UniqueROBuffer) -> &'static mut [u8] { buffer.buffer }
}

impl core::ops::Deref for UniqueROBuffer {
    type Target = [u8];
    fn deref(&self) -> &[u8] { &self.buffer }
}
```

Reply:
The only benefit of `UniqueROBuffer` over `&'static mut [u8]` is it provides immutable rather than mutable access, so it might not be worth the cognitive overhead the extra type and conversions.

Reply:
This sounds somewhat similar to `LeasableBuffer` (https://github.com/tock/tock/blob/master/kernel/src/common/leasable_buffer.rs), which is used for passing a portion of a buffer to a lower-layer call while still getting the entire buffer back in the callback.

There was not really enough momentum to motivate converting existing code that could have used this, so it remains mostly used by the networking stack,

Reply:
> Musing: I wonder if we should have some sort of `UniqueROBuffer` that provides immutable access to a reference, but which is guaranteed to be the only borrow of a given buffer. Something like:

As Hudson pointed out, parts of Tock have tried doing something like this where it was particularly important/useful. I think types such as these aren't incompatible with this rule. You're no longer passing a buffer, but rather a type that provides more refined access semantics.